---
title: "Homework 2"
author: "Almog Angel"
Students: "<Lev Tsipris> <317600492> <Liad> <316225523>"
date: "15/02/2024"
output: html_document
---

- Load packages and set working directory
```{r}
BiocManager::install("DESeq2")
BiocManager::install("MatrixGenerics")
library(DESeq2)

install.packages("devtools")
library(devtools)

install.packages("ashr")
library(ashr)

BiocManager::install("EnhancedVolcano")
library(EnhancedVolcano)

BiocManager::install("factoextra")
library(factoextra)

BiocManager::install("clusterProfiler")
library(clusterProfiler)

BiocManager::install("msigdbr")
library(msigdbr)

install.packages("tidyverse")
library(tidyverse)

dir <- getwd()
setwd(dir)
```

- In this homework assignment, you will perform differential gene expression analysis
- Read the abstract of the study "Temporal changes in postprandial blood transcriptomes reveal subject-specific pattern of expression of innate immunity genes after a high-fat meal" by Danielle G Lemay et a. (pdf file attached)

(1) Explain in your own words the term "whole blood transcriptome" (up to 30 words)
(2) Explain in your own words the objective and design of the study (up to 250 words)
```{}
Write your answer here:
(1) Genome-wide expression (RNA) of all genes from all cells which are present in the blood (WBC, RBC & platelets).

(2) The paper is actually a sub-study of a study which investigate the effect of dietry supplaments on the immune response after a
    fat-rich meal.
     
    The objective of the sub-study is to determine which genes are transcripted in white blood cells, after consuming a high-fat meal.
    Specifically, the study looks into the the temporal change and subject variabality in expression of innate immunity genes.
    
    The blood for analysis is drawn at 3 diffrent times: after 12h fasting before eating, 3 hours after meal and 6 hours after meal.
    There are 5 subjects, each given 3 meals on different days, each with different dietry supplaments which had an insegnificant effect 
    Thus overall 3x5x3=45 transcriptomes where performed.
    The meals and fasting times of the subjects were designed to minimize uncontrolled dietry effects.

    The blood was shaken, frozen, then total RNA isolated and enriched for non-RBC and non-ribosomal RNA, then sequenced.
    WBC transcriptome was analyzed using various statistical methods for differential expression across experimental factors: Subject, Time, Arm                 (dietry supplaments added), Day.
    The transcriptome was also analyzed for enrichment in immune response pathways.
    
    The target genes were validated using RT-PCR in 180 samples (from 20 subjects).
    
```

- Take a look at the manuscript, find the Gene Expression Omnibus (GEO) accession number
- Write it down here:
```{}
GEO accession number: study accession number GSE127530 and series record GSE127530
```

- Go in to the GEO website and search for this accession number.
- Download the counts data (the second one: "fixed")
- Download the Series Matrix File(s)

- Read the count matrix and metadata into the corresponding variables below
```{r}

library(readr)
counts <- read.table('GSE127530_fixed_combinedCounts.txt.gz', stringsAsFactors = FALSE, sep='\t', header=T, check.names = FALSE)

packages_for_tutorial <- c("DESeq2", "ggplot2", "GEOquery", "AnnotationDbi", "org.Hs.eg.db", "EnhancedVolcano", "pheatmap",
                           "factoextra", "gridExtra", "corrplot", "dendextend")

not_installed <- packages_for_tutorial[!(packages_for_tutorial %in% installed.packages()[ , "Package"])]

for (package in not_installed) {
  BiocManager::install(package) 
}

library(GEOquery)
library(AnnotationDbi)
library(org.Hs.eg.db)

metadata <- getGEO(filename='GSE127530_series_matrix.txt')
```

- Convert metadata to a tibble
- Select and rename the following columns and store it back into "metadata":
Sample = description
Subject = `study subject id #:ch1`
Day = `test day:ch1`
Time = `time of blood draw:ch1`
- Make sure the "Time" column is in factor class
```{r}
library(dplyr)
metadata_tibble <- as_tibble(metadata, .name_repair="minimal")

metadata <- metadata_tibble %>% 
  dplyr::select(Sample = 'description', Subject = `study.subject.id...ch1`,
                 Day = `test.day.ch1`, Time = `time.of.blood.draw.ch1`)
 metadata <- mutate(metadata, Time = as.factor(Time))
 metadata
```

- Make the reference level for Time be the "Fasting" condition
```{r}
metadata$Time <- relevel(metadata$Time, ref="fasting")
```

- Remove genes with zero total counts
```{r}
counts <- counts[rowSums(counts)>0,]
```


- Make a DESeq2 object called "dds" and use the Time column for the design
* Don't forget to check that samples in colData and countData match
* In this example tidy = FALSE (use ?DESeqDataSetFromMatrix to learn why)
```{r}
library(DESeq2)

counts <- counts[metadata$Sample]
all(metadata$Sample == colnames(counts))

dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = metadata,
                              design= ~ Time, tidy = FALSE)

```

- Run DESeq2
```{r}
dds <- DESeq(dds)
```

- We are now comparing the groups of fasting and time = 3 hours 

- Use the function lfcShrink() to adjust the LFC values with the following arguments:
  (1) coef="Time_3.hr.postprandial_vs_fasting" for the name of the coefficient to shrink
  (2) type="apeglm"
- Store the results back into "res"
```{r}
BiocManager::install("apeglm")

res <- lfcShrink(dds, coef="Time_3.hr.postprandial_vs_fasting", type="apeglm")
```
- Sort "res" by the adjusted p-values from lowest to highest and store in "resOrdered"
```{r}
resOrdered <- res[order(res$padj),]
resSig<-(resOrdered$padj<0.05)
max(which(resSig == TRUE))
```

- Did you find any significant genes (adjusted p-value < 0.05)? If so, how many?
```{}
Write your answer here:
85 significant genes
```

- Use the function vst() to extract the normalized counts from dds into "counts.vst"
```{r}
?vst()
counts.vst <- vst(dds)
```

- Use the function plotPCA() to generate a ggplot2 object for the PCA visualization and store it into "pcaData":
  (1) Use the top 1000 variable genes
  (2) Use intgroup=c("Subject", "Day") for grouping
  (3) Use returnData = TRUE
```{r}
pcaData <- plotPCA(counts.vst, intgroup=c("Subject", "Day"), ntop = 1000, returnData = TRUE) 
```

- Use ggplot to plot the object you made above:
  (1) Plot the percent of variance explained in PC1 and PC2 labels
  (2) Color points by Subject
  (3) Shape points by Day
```{r}
# install.packages("ggplot2")
library(ggplot2)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(x = PC1, y = PC2, color = Subject, shape = Day)) +
  geom_point(size =3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  ggtitle("PCA with VST data")
```


- Write a short script that preform hierarchical or kmeans clustering (choose one):
(1) Extract the data from "counts.vst" using the assay(function)
(2) Use only top 1,000 variable genes
(3) Calculate distance matrix
(4) Use the Elbow method to decide the optimal number of clusters
(5) Plot the results in a dendogram for the hierarchical clustering or use fviz_cluster() for kmeans

```{r}
normcounts <- assay(counts.vst) 
var_per_gene <- apply(normcounts, 1, var) # Calculate the variance per gene
selectedGenes <- names(var_per_gene[order(var_per_gene, decreasing = T)][1:1000]) # Take the top 1000 variable genes
normcounts.top1Kvar <- t(normcounts[selectedGenes,]) # Construct a new matrix only for the top 1000 genes
# Construct a distance matrix based on pairwise euclidean distances between all counts 
dist_mat <- dist(normcounts.top1Kvar, method = 'euclidean')
hclust_avg <- hclust(dist_mat, method = 'average')
plot(hclust_avg, cex = 0.6, hang = -1)
library(factoextra)
fviz_nbclust(normcounts.top1Kvar, FUN = hcut, method = "wss")
# optimal is k=5 clusters

```


- Looking at the PCA and the clustering you made above:
(1) What is the main source of variation among samples in this study?
(2) What is the second source of variation?
(3) What led you to this conclusion?
```{}
Write your answer here:
(1) the subject
(2) the day
(3) subjects create well seperated clusters in PCA & dendogram. 
    in figure 1A time is a worse sperating varibale than day in our PCA plot.
```
* Hint: Look at Figure 1A in the manuscript.


- Make a second DESeq2 object called "dds2"
- This time make use of the information you gained from the PCA above to in the "design" argument
```{r}
dds2 <- DESeqDataSetFromMatrix(countData = counts,
                              colData = metadata,
                              design= ~ Time+Subject+Day, tidy = FALSE)
```

- How did your new object differ from the old one? Explain why your results should be better now.
```{}
Write your answer here:
We now test for additional and more primary sources of variation (primarly the subject, and also day), in contrast to the first analysis.
The results whould improve as now our model includes more variance sources and hence will detect more variance in the RNA expression. 
```

- Run DESeq() and lfcShrink() again, save the results into "res2" and sort by adjusted p-value in "resOrdered2"
```{r}
dds2 <- DESeq(dds2)
resultsNames(dds2) 
res2 <- lfcShrink(dds2, coef="Time_3.hr.postprandial_vs_fasting", type="apeglm")
resOrdered2 <- res2[order(res2$padj),]
resSig2<-(resOrdered2$padj<0.05)
max(which(resSig2 == TRUE))
```

- How many significant genes (adjusted p-value < 0.05) did you get this time?
```{}
Write your answer here:
2154 significant genes
```

- Explain in short the meaning of each column in "resOrdered2"
```{}
Write your answer here:
baseMean - The average of the normalized gene count values, dividing by size factors, taken over all samples.
log2FoldChange - The effect size estimate. This value indicates how much the gene or transcript's expression seems to have changed between the comparison and control groups. This                  value is reported on a logarithmic scale to base 2.
lfcSE - The standard error estimate for the log2 fold change estimate
pvalue - P-value of the test for the gene or transcript
padj - Adjusted P-value for multiple testing for the gene or transcript
```

- Visualize the results with a volcano plot as we did in the tutorial
- Change the following arguments to:
  (1) FCcutoff=0.5
  (2) pCutoff=0.05
  (3) xlim=c(-1.5, 1.5)
  (4) ylim=c(0, 30)
```{r}
library(EnhancedVolcano)
EnhancedVolcano(resOrdered2,
                lab = rownames(resOrdered2),
                x = 'log2FoldChange',
                y = 'padj',
                labSize=3,
                FCcutoff=0.5,
                pCutoff=0.05,
                xlim=c(-1.5, 1.5),
                ylim=c(0, 30),
                )

```

- Look at the volcano plot and and explain in short the meaning of each axis:
```{}
log2FoldChange (x-axis) - The effect size estimate. This value indicates how much the gene or transcript's expression seems to have changed between the comparison and control groups. This value is reported on a logarithmic scale to base 2 (how strongly the gene is up/down transcripted)
padj (y-axis) - Adjusted P-value for multiple testing for the gene or transcript (how significant the gene expression is)
```


- Now lets try to understand the meaning of the differential expressed genes (DEG).
- Take a look at the genes names (rownames)
```{r}
View(data.frame(resOrdered2))
```

- Pick several genes you like from the most DEG
- Google their names, try to find if they have anything in common
- In your answer, mention the genes and what did you found about them
```{}
Write your answer here:

PER1: Period Circadian Regulator
DDIT4: Involved in defense response to virus
SLC25A20: This protein is essential for fatty acid oxidation

We can see the genes are related to for example time passing / fats digetsion,
which makes sense for the 3-hour before/after meal comparison we are performing.

```

- Now you  will use functional enrichment analysis with the Hallmark pathways gene sets.
- We learned two ways to perform functional enrichment analysis - over representation and gene set enrichment analysis (GSEA). Here you will run GSEA.  

- First we need to create an ordered vector by the log fold change with the gene symbols as names:
```{r}
resOrdered2.fatDiet.nona <- resOrdered2[!is.na(resOrdered2$padj) & resOrdered2$log2FoldChange > 0,]
genes_ordered <- sort(resOrdered2.fatDiet.nona$log2FoldChange, decreasing = T)
```

- We now need to get the Hallmarks pathways gene sets. We will use the msigdbr package for that:
```{r}
library(msigdbr)
hallmarks <- msigdbr(species = "Homo sapiens", category = "H") %>% 
  dplyr::select(gs_name, gene_symbol)
```

- Use the GSEA() function from the clusterProfiler package to run the analysis.
- You can find an awesome tutorial for the clusterProfiler package here: http://yulab-smu.top/biomedical-knowledge-mining-book/universal-api.html
- Save your results into "hm"
```{r}
library(clusterProfiler)
hm <- GSEA(genes_ordered, TERM2GENE=hallmarks, scoreType = "pos")
```

- Finally, visualize the results of this analysis using the dotPlot function from the clusterProfiler database.
```{r}
hallmarks <- msigdbr(species = "Homo sapiens", category = "H")
hallmarks <- hallmarks[,c('gs_name', 'gene_symbol')]
dotplot(hm)
```
- Summarize your conclusions from the analysis (up to 250 words)
- Make sure you base your conclusions on the different steps you made (PCA, clustering, DEG, GSEA) 
```{}
Write your answer here:

Looking at our PCA clusters, we can see that pre and post fatty meal gene transcription variability is mostly explained by the difference between subjects.
There is also less strong clustring with respect to the day variable of the experiment, that can be seen in the clustering dendogram.
Some of the differentially expressed genes are related to expected function after having a fatty meal or to simply time passing, as we can see in the DEG.
Looking at the GSEA, there is expression of immune related pathways such as the inflammatory response IL-6/JAK2/STAT3.
Thus we can confidently link fatty meal consupmtion with some sort of immune response.

```

- Save this HW as HTML  

Eat healthy :)